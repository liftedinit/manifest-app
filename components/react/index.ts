export * from './addressCopy';
export * from './modal';
export * from './views';
export * from './settingsModal';
export * from './inputs';
export { SignModal } from './authSignerModal';
export * from './ModalDialog';
