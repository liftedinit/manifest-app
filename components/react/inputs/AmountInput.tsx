import BigNumber from 'bignumber.js';
import React from 'react';

export interface AmountInputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  value: string | number | undefined;
  onValueChange: (newAmount: BigNumber | undefined) => void;
}

/**
 * A component for entering an amount of tokens. Allows only positive decimal
 * numbers, or empty values (for an empty field).
 * @param value The current value of the input field.
 * @param onValueChange A callback that is called when the value of the input field
 *                      changes, with the new amount.
 * @param className Additional classes to apply to the input field.
 * @param props Additional props to pass to the input field.
 * @constructor
 */
export const AmountInput: React.FC<AmountInputProps> = ({
  value,
  onValueChange,
  className,
  ...props
}) => {
  const [internalValue, setInternalValue] = React.useState<string>(
    value === undefined ? '' : value.toString()
  );

  const [lastValue, setLastValue] = React.useState<string | number | undefined>(value);

  React.useEffect(() => {
    if (value !== lastValue) {
      setInternalValue(value === undefined ? '' : value.toString());
      setLastValue(value);
    }
  }, [value, lastValue]);

  function onChange(event: React.ChangeEvent<HTMLInputElement>) {
    const v = event.target.value;

    // Allow for `.` to be entered on its way to a real number.
    if (v === '' || v === '.') {
      setInternalValue(v);
      onValueChange(undefined);
      return;
    }

    const newValue = /^\d*\.?\d*$/.test(v) ? new BigNumber(v) : null;

    if (newValue !== null) {
      setInternalValue(v);
      onValueChange(newValue);
    } else {
      onValueChange(internalValue === '' ? undefined : new BigNumber(internalValue));
    }
  }

  return (
    <input
      className={`input input-md border border-[#00000033] dark:border-[#FFFFFF33] bg-[#E0E0FF0A] dark:bg-[#E0E0FF0A] w-full pr-24 dark:text-[#FFFFFF] text-[#161616] rounded-xl ${className}`}
      type="text"
      placeholder="0.00"
      min={0}
      value={internalValue}
      onChange={onChange}
      onKeyDown={e => {
        if (e.key.length === 1 && !/[\d.]/.test(e.key) && !e.ctrlKey && !e.metaKey && !e.altKey) {
          e.preventDefault();
        }
      }}
      {...props}
    />
  );
};
