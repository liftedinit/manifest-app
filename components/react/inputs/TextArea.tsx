import { useField } from 'formik';
import React from 'react';

interface TextAreaProps {
  label: string;
  name: string;
  className?: string;
}

export const TextArea: React.FC<
  TextAreaProps & React.TextareaHTMLAttributes<HTMLTextAreaElement>
> = ({ label, ...props }) => {
  const [field, meta] = useField(props);
  const id = props.id || props.name;
  return (
    <div className="form-control w-full">
      <label className="label" htmlFor={id}>
        <span className="label-text text-[#00000099] dark:text-[#FFFFFF99]">{label}</span>
      </label>
      <div className="relative">
        <textarea
          id={id}
          {...field}
          {...props}
          className={`dark:text-[#FFFFFF99] text-[#161616] textarea border-[#00000033] dark:border-[#FFFFFF33] bg-[#E0E0FF0A] dark:bg-[#E0E0FF0A] w-full 
            autofill:bg-[#E0E0FF0A] dark:autofill:bg-[#E0E0FF0A]
            focus:bg-[#E0E0FF0A] dark:focus:bg-[#E0E0FF0A]
            ${props.className}`}
        />
      </div>
      {meta.touched && meta.error ? (
        <label className="label">
          <span className="label-text-alt text-error">{meta.error}</span>
        </label>
      ) : null}
    </div>
  );
};
