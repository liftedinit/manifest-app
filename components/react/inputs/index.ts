export * from './AmountInput';
export * from './TextInput';
export * from './NumberInput';
export * from './TextArea';
