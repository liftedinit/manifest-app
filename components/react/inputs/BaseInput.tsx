import React from 'react';
import { useField } from 'formik';
import { QuestionIcon } from '@/components/icons';
interface BaseInputProps {
  label?: string;
  name: string;
  className?: string;
  rightElement?: React.ReactNode;
  showError?: boolean;
  leftElement?: React.ReactNode;
  helperText?: string;
}

export const BaseInput: React.FC<BaseInputProps & React.InputHTMLAttributes<HTMLInputElement>> = ({
  label,
  rightElement,
  leftElement,
  helperText,
  showError = true,
  ...props
}) => {
  const [field, meta] = useField(props);
  const id = props.id || props.name;
  return (
    <div className="form-control w-full">
      <div className="flex justify-between items-center">
        {label && (
          <label className="label" htmlFor={id}>
            <span className="label-text text-[#00000099] dark:text-[#FFFFFF99] select-text">
              {label}
            </span>
          </label>
        )}
        {helperText && (
          <span className="text-sm flex-row flex items-center justify-between gap-2 text-[#00000099] dark:text-[#FFFFFF99]">
            <QuestionIcon className="w-4 h-4" /> {helperText}
          </span>
        )}
      </div>
      <div className="relative">
        <input
          id={id}
          {...field}
          {...props}
          className={`dark:text-[#FFFFFF99] text-[#161616] input border-[#00000033] dark:border-[#FFFFFF33] bg-[#E0E0FF0A] dark:bg-[#E0E0FF0A] w-full 
            autofill:bg-[#E0E0FF0A] autofill:dark:bg-[#E0E0FF0A]
            focus:bg-[#E0E0FF0A] focus:dark:bg-[#E0E0FF0A]
            ${props.className}`}
        />
        {rightElement && (
          <div className="absolute inset-y-0 right-0 flex items-center pr-3">{rightElement}</div>
        )}
        {leftElement && (
          <div className="absolute inset-y-0 left-0 flex items-center pl-3">{leftElement}</div>
        )}
      </div>
      {meta.touched && meta.error && showError ? (
        <label className="label">
          <span className="label-text-alt text-error">{meta.error}</span>
        </label>
      ) : null}
    </div>
  );
};
