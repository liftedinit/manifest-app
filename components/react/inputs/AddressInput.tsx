import { useChain } from '@cosmos-kit/react';
import { Dialog } from '@headlessui/react';
import React from 'react';
import { MdContacts } from 'react-icons/md';

import { Contacts, TextInput } from '@/components';
import { BaseInputProps } from '@/components/react/inputs/BaseInput';

import env from '../../../config/env';

export interface AddressInputProps extends BaseInputProps {
  rightElement?: React.ReactNode;
  name: string;
  small?: boolean;
}

export const AddressInput: React.FC<AddressInputProps> = ({
  value: initialValue,
  name,
  onChange,
  rightElement,
  disabled,
  small,
  ...props
}) => {
  const [value, setValue] = React.useState(initialValue);
  const [showContacts, setShowContacts] = React.useState(false);
  const { address } = useChain(env.chain);

  return (
    <>
      <TextInput
        onChange={e => {
          setValue(e.target.value);
          onChange?.(e);
        }}
        name={name}
        value={value}
        disabled={disabled}
        rightElement={
          <div className="flex gap-2">
            {!disabled && (
              <button
                type="button"
                onClick={() => setShowContacts(true)}
                className={`btn btn-primary text-white ${small ? 'btn-xs' : 'btn-sm'}`}
              >
                <MdContacts className={small ? 'w-4 h-4' : 'w-5 h-5'} />
              </button>
            )}
            {rightElement}
          </div>
        }
        {...props}
      />

      {!disabled && showContacts && (
        <Dialog
          className={`modal modal-open fixed flex p-0 m-0 z-1`}
          style={{
            height: '100vh',
            width: '100vw',
            alignItems: 'center',
            justifyContent: 'center',
          }}
          onClose={() => setShowContacts(false)}
          open={showContacts}
        >
          <div className="fixed inset-0 bg-black/30" aria-hidden="true" />
          <Dialog.Panel className="modal-box bg-secondary rounded-[24px] p-6">
            <Contacts
              onClose={() => setShowContacts(false)}
              currentAddress={address ?? ''}
              onSelect={address => {
                setValue(address);
                onChange?.({ target: { name, value: address } } as any);
              }}
              selectionMode
            />
          </Dialog.Panel>
        </Dialog>
      )}
    </>
  );
};
