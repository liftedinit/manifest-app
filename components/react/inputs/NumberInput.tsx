import React from 'react';

import { BaseInput } from './BaseInput';

export const NumberInput: React.FC<React.ComponentProps<typeof BaseInput>> = props => (
  <BaseInput type="number" {...props} />
);
