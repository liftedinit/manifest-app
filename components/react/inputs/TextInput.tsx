import React from 'react';
import { BaseInput } from './BaseInput';

export const TextInput: React.FC<React.ComponentProps<typeof BaseInput>> = props => (
  <BaseInput type="text" data-1p-ignore {...props} />
);
