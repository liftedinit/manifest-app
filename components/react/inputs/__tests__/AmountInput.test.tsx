import { jest, test, expect, afterEach, describe } from 'bun:test';
import React from 'react';
import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import { AmountInput, NumberInput } from '@/components/react/inputs';
import { Formik, Form } from 'formik';

const TestForm = ({ children }: { children: React.ReactNode }) => (
  <Formik initialValues={{ test: '' }} onSubmit={() => {}}>
    <Form>{children}</Form>
  </Formik>
);

describe('AmountInput', () => {
  afterEach(() => {
    cleanup();
  });

  test('renders correctly', () => {
    const value = '42';
    const onValueChange = jest.fn();

    render(
      <TestForm>
        <AmountInput name="test" value={value} onValueChange={onValueChange} />
      </TestForm>
    );

    const input = screen.getByPlaceholderText('0.00');
    expect(input).toBeInTheDocument();
    expect(input.tagName.toLowerCase()).toBe('input');
    expect(input).toHaveAttribute('type', 'text');
  });

  test('calls onValueChange with the new value', () => {
    const value = '42';
    const onValueChange = jest.fn();

    render(
      <TestForm>
        <AmountInput name="test" value={value} onValueChange={onValueChange} />
      </TestForm>
    );

    const input = screen.getByPlaceholderText('0.00');
    fireEvent.change(input, { target: { value: '42.42' } });
    expect(onValueChange).toHaveBeenCalledWith('42.42');
  });

  test('calls onValueChange with an empty string when the input is empty', () => {
    const value = '42';
    const onValueChange = jest.fn();

    render(
      <TestForm>
        <AmountInput name="test" value={value} onValueChange={onValueChange} />
      </TestForm>
    );

    const input = screen.getByPlaceholderText('0.00');
    fireEvent.change(input, { target: { value: '' } });
    expect(onValueChange).toHaveBeenCalledWith('');
  });

  test('calls onValueChange with the same value when the input is invalid', () => {
    const value = '42';
    const onValueChange = jest.fn();

    render(
      <TestForm>
        <AmountInput name="test" value={value} onValueChange={onValueChange} />
      </TestForm>
    );

    const input = screen.getByPlaceholderText('0.00');
    fireEvent.change(input, { target: { value: '42.42.42' } });
    expect(onValueChange).toHaveBeenCalledWith('42');
  });

  test('calls onValueChange with the same value when the input is invalid (empty initial value)', () => {
    const value = '';
    const onValueChange = jest.fn();

    render(
      <TestForm>
        <AmountInput name="test" value={value} onValueChange={onValueChange} />
      </TestForm>
    );

    const input = screen.getByPlaceholderText('0.00');
    fireEvent.change(input, { target: { value: '42.42.42' } });
    expect(onValueChange).toHaveBeenCalledWith('');
    onValueChange.mockClear();
    fireEvent.change(input, { target: { value: '1' } });
    expect(onValueChange).toHaveBeenCalledWith('1');
    onValueChange.mockClear();
    fireEvent.change(input, { target: { value: '1E+10' } });
    expect(onValueChange).toHaveBeenCalledWith('');
  });

  test('calls onValueChange when only a dot is present', () => {
    const value = '42';
    const onValueChange = jest.fn();

    render(
      <TestForm>
        <AmountInput name="test" value={value} onValueChange={onValueChange} />
      </TestForm>
    );

    const input = screen.getByPlaceholderText('0.00');
    fireEvent.change(input, { target: { value: '.' } });
    expect(onValueChange).toHaveBeenCalledWith('.');
  });
});
