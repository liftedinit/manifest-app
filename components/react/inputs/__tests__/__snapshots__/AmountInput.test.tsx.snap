// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountInput renders correctly 1`] = `
"<DocumentFragment>
  bound HTMLFormElement {
    "0": <input
      class="input input-md border border-[#00000033] dark:border-[#FFFFFF33] bg-[#E0E0FF0A] dark:bg-[#E0E0FF0A] w-full pr-24 dark:text-[#FFFFFF] text-[#161616] rounded-xl undefined"
      min="0"
      name="test"
      placeholder="0.00"
      type="text"
      value="42"
    />,
  }
</DocumentFragment>"
`;
